File: day03/solution.js
import fs from 'fs/promises';
import process from 'node:process';

// prettier-ignore
const priorities = [
  'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z',
  'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z'
]

const input = await fs.readFile('./input.txt', 'utf8');
const lines = input.split('\n');

let sum = 0;
if (process.env.part === 'part2') {
  const packs = ['', '', ''];
  let pack_count = 0;
  for (const line of lines) {
    packs[pack_count] = line;
    pack_count += 1;
    if (pack_count === 3) {
      pack_count = 0;
      outer: for (const a of packs[0]) {
        for (const b of packs[1]) {
          for (const c of packs[2]) {
            if (a === b && b === c) {
              sum += priorities.indexOf(a) + 1;
              break outer;
            }
          }
        }
      }
    }
  }
} else {
  for (const line of lines) {
    const pack_size = line.length / 2;
    const first = line.slice(0, pack_size);
    const second = line.slice(pack_size);
    outer: for (const a of first) {
      for (const b of second) {
        if (a === b) {
          sum += priorities.indexOf(a) + 1;
          break outer;
        }
      }
    }
  }
}
console.log(sum);

File: day03/solution.nim
import system/io
import std/os

const priorities = [
  'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z',
  'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z'
]

var sum = 0

if getEnv("part") == "part2":
  var packs = ["", "", ""]
  var pack_count = 0
  for line in lines "input.txt":
    packs[pack_count] = line
    pack_count += 1
    if pack_count == 3:
      pack_count = 0
      block outer:
        for a in packs[0]:
          for b in packs[1]:
            for c in packs[2]:
              if a == b and b == c:
                sum += find(priorities, a) + 1
                break outer
else:
  for line in lines "input.txt":
    var pack_size = (line.len / 2).int
    var first = line[0..pack_size-1]
    var second = line[pack_size..line.len-1]
    block outer:
      for a in first:
        for b in second:
          if a == b:
            sum += find(priorities, a) + 1
            break outer

echo sum
